gopher/normalize.py

```diff
@@ -0,0 +1,82 @@
+import pandas as pd
+from Bio import SeqIO
+
+
+def normalize(proteins, fasta):
+    """Normalize intensity values.
+    Parameters
+    ----------
+    proteins : pandas.DataFrame
+        A dataframe where the indices are UniProt accessions and each column is
+        an experiment to test. The values in this dataframe should be some
+        measure of protein abundance: these could be the raw measurement if
+        originating from a single sample or a fold-change/p-value if looking at
+        the difference between two conditions.
+    fasta : fasta file
+        Use the FASTA file to generate molecular weights for normalization
+    Returns
+    -------
+    pandas.DataFrame
+        The normalized intensities for every protein in each sample.
+    """
+    fasta_df = pd.DataFrame(columns=["Protein", "Sequence", "Mass"])
+    fasta_sequences = SeqIO.parse(open(fasta), "fasta")
+    for f in fasta_sequences:
+        name = f.id.split("|")[1]
+        mass = calculate_mass(f.seq)
+        temp = {"Protein": name, "Sequence": str(f.seq), "Mass": mass}
+        temp = pd.DataFrame(temp, index=[0])
+        fasta_df = pd.concat([fasta_df, pd.DataFrame(temp)], ignore_index=True)
+
+    proteins = proteins.apply(col_norm, axis=0)
+    df = pd.merge(fasta_df, proteins, on="Protein")
+    df = df.set_index("Protein")
+    df = df.drop(columns=["Sequence"])
+    df = df.apply(mass_norm, axis=1)
+    df = df.drop(columns=["Mass"])
+
+    return df
+
+
+#
+def col_norm(col):
+    """Calculate the ratio of protein/total protein"""
+    return col / col.sum()
+
+
+def mass_norm(row):
+    """Do mass part of normalization"""
+    mass = row.loc["Mass"]
+    return row / mass
+
+
+def calculate_mass(seq):
```

Comment:
Since we're already using Biopython, I use their function: https://biopython.org/docs/1.75/api/Bio.SeqUtils.html#Bio.SeqUtils.molecular_weight

Reply:
Thanks! That is much better.